contracts/bridge/src/fee-market/FeeMarket.sol

```diff
@@ -385,70 +377,131 @@ contract FeeMarket is IFeeMarket {
         _reward(VAULT, total_vault_reward);
     }
 
-    function _reward_and_unlock_ontime(
+    function _settle_order(
         uint256 key,
-        uint256 diff_time,
         address delivery_relayer,
         address confirm_relayer
-    ) private returns (uint256 delivery_reward, uint256 confirm_reward, uint256 vault_reward) {
-        // get the message fee from the last top N relayers
+    ) private returns (
+        uint256 delivery_reward,
+        uint256 confirm_reward,
+        uint256 vault_reward
+    ) {
+        Order memory order = orderOf[key];
+        require(orderOf[key].assignedTime > 0, "!exist");
+        // Get the message fee from the top N relayers
         uint256 message_fee = getOrderFee(key);
+        // Get slot index and slot price
+        (uint256 slot, uint256 slot_price) = _get_slot_price(key, message_fee);
+        // Message surplus = Message Fee - Slot price
+        uint256 message_surplus = message_fee - slot_price;
+        // A. Slot Offensive Slash
+        uint256 slot_offensive_slash = _do_slot_offensive_slash(key, slot);
+        // Message Reward = Slot price + Slot Offensive Slash
+        uint256 message_reward = slot_price + slot_offensive_slash;
+        // B. Slot Duty Reward
+        uint256 slot_duty_reward = _do_slot_duty_reward(key, slot, message_surplus);
+        // Message Reward -> (delivery_relayer, confirm_relayer)
+        (delivery_reward, confirm_reward) = _distribute_fee(message_reward);
+        // Message surplus -= Slot Duty Reward
+        require(message_surplus >= slot_duty_reward, "!surplus");
+        vault_reward = message_surplus - slot_duty_reward;
+    }
+
+    function _get_order_status(
+        uint key
+    ) private view returns (
+        bool is_ontime,
+        uint256 diff_time,
+        uint256 number,
+        uint256 collateral
+    ) {
         Order memory order = orderOf[key];
-        for (uint slot = 0; slot < order.assignedRelayersNumber; slot++) {
-            address assign_relayer = assignedRelayers[key][slot].assignedRelayer;
-            // the message delivery in the `slot` assign_relayer
-            // [slot, slot+1)
-            if (slot * relayTime <= diff_time && diff_time < (slot + 1 ) * relayTime) {
-                uint256 base_fee = assignedRelayers[key][slot].makerFee;
-                (delivery_reward, confirm_reward, vault_reward) = _distribute_ontime(message_fee, base_fee, assign_relayer, delivery_relayer, confirm_relayer);
+        number = order.assignedRelayersNumber;
+        collateral = order.collateral;
+        // Diff_time = settle_time - assign_time
+        diff_time = block.timestamp - order.assignedTime;
```

Comment:
overflow check?

Reply:
`settle_time` > `assign_time` in the same order.